File: planning/ensure_labels.rb
require 'json'
require 'open3'

bundle = JSON.parse(File.read(File.join(__dir__, 'backlog.json')))
repo = bundle.fetch('repository')
output, error, status = Open3.capture3('gh', 'api', "repos/#{repo}/labels?per_page=100")
abort error unless status.success?
existing = JSON.parse(output).map { |label| label.fetch('name') }
created = []
bundle.fetch('labels').each do |name|
  next if existing.include?(name)
  color = name.start_with?('wave-') ? 'bfd4f2' : name.start_with?('domain:') ? 'c2e0c6' : 'd4c5f9'
  description = name == 'blocked' ? 'Not dispatchable: dependency, approval, or explicit dispatch hold' : 'Tradvisor V1 architecture delivery classification'
  output, error, status = Open3.capture3('gh', 'api', '--method', 'POST', "repos/#{repo}/labels", '-f', "name=#{name}", '-f', "color=#{color}", '-f', "description=#{description}")
  unless status.success?
    abort "STOP: failed label #{name}; created #{created.join(', ')}; reconcile before retry. #{error}"
  end
  result = JSON.parse(output)
  abort "STOP: unexpected label response for #{name}" unless result['name'] == name
  created << name
end
puts "Verified label setup; created #{created.length}, preserved #{existing.length} existing labels."

File: planning/ensure_labels.py
#!/usr/bin/env python3
import json, subprocess
from pathlib import Path
B = json.loads((Path(__file__).parent/'backlog.json').read_text())
repo=B['repository']; existing=json.loads(subprocess.check_output(['gh','api',f'repos/{repo}/labels?per_page=100']))
names={x['name'] for x in existing}; created=[]
for name in B['labels']:
    if name in names: continue
    color='bfd4f2' if name.startswith('wave-') else ('c2e0c6' if name.startswith('domain:') else 'd4c5f9')
    desc='Not dispatchable: dependency, approval, or explicit dispatch hold' if name=='blocked' else 'Tradvisor V1 architecture delivery classification'
    subprocess.check_call(['gh','api','--method','POST',f'repos/{repo}/labels','-f',f'name={name}','-f',f'color={color}','-f',f'description={desc}'])
    created.append(name)
print(f'Verified label setup; created {len(created)}, preserved {len(names)} existing labels.')
